### PR TITLE
Readying linear article

### PR DESCRIPTION
**Pull request by Verdagon**

### CODE REVIEW

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+The other two languages say the same thing as each other, and also mention those same kings and queens.
+
+
+The scholar Jean-Francois Champollion used these clues to eventually decipher the entire stone, and enabled us to finally understand ancient Egyptian hieroglyphs.
+
+
+I wonder what Champollion felt in that moment, when he had that entire slate translated for the first time.
+
+
+I'm no scholar, and I certainly didn't discover anything new. But I think I felt like Champollion did, when Vale showed me the hidden truth about how we can use linear types to completely eliminate Vale's memory safety costs, and led to the creation of the *linear-aliasing model*.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: Very long sentence, hard to parse. Suggest breaking it up:

I'm no scholar and linear types are not anything new, but I still felt a glimmer of what Champollion felt when Vale unearthed clues connecting linear types to memory safety. What hidden truth was revealed? We can use linear types to completely eliminate Vale's memory safety costs, a new paradigm called the *linear-aliasing model*!

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+
+
+<slice/>
+
+# A different way to get memory safety
+
+Recently, I was making a sample program to show how Vale's [regions](/blog/zero-cost-borrowing-regions-overview) can blend borrowing with shared mutability.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: Will everyone know what "borrowing" is here?

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+
+
+I was surprised to find that the sample program had almost _zero_ memory-safety overhead, except for a line in this ([paraphrased](https://github.com/Verdagon/RegionsBenchmarks/blob/main/cellular-automata/CellularAutomata.vale)) code:
+
+[# E: These examples clear enough? Trying to balance noisiness and explanation, so was sparse on comments.]
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: Yep, should you comment that x = rand.num; is the line in question?

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+```
+
+
+Specifically, when `RandInt`'s `rand.num` dereferences the `rand` object, it does an implicit [generation check](/blog/generational-references) assertion to ensure the object is still alive.
+
+
+
+[# E: This paragraph is here because I'm trying to convey that generation checks _aren't_ expensive, and not really worth optimizing away much of the time, but we *can* if we want to.]
+
+Normally, I don't believe in chasing zero-cost memory safety. Playing overhead golf can often lead to chasing diminishing returns or prematurely optimizing. After all, a generation check is usually only expensive as a bounds check (in other words, usually negligible).
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: as

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+Here's what we changed:
+
+ * `rand &Random` became `rand Random`. Instead of taking a reference, the caller moves the actual Random instance itself.
+ * `return x` became `return (x, rand);` which returns a two-element tuple, which also has the Random instance.
+ * The return type `i64` became `(i64, Random)` accordingly.
+
+
+Now, `rand.num` isn't dereferencing a reference, it's accessing a field of an owned value.
+
+Vale's compiler sees this, and doesn't insert a generation check, because it knows the object is alive.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: consider adding "since we own it", feel free to ignore if you think this is self-evident enough

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+[r, set rand] = RandInt(rand);
+```
+
+where `rand` is an actual `Random`, not just a reference.
+>>>>
+
+
+Now, that line:
+
+ * Moves the `rand` instance into the `RandInt` call.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:165]: Consider "function's memory space", depending how pedantic you want this article to be

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+func MakeBoard(
+    rand &Random,
+    num_rows int,
+    num_cols int)
+[][]bool {
+  rows = [][]bool(num_rows);
+  foreach row_i in 0..num_rows {
+    row = []bool(num_cols);
+    foreach col_i in 0..num_cols {
+      r = RandInt(rand);
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:188]: Any chance you can bold this?

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+func MakeBoard(
+   rand Random,
+   num_rows int,
+   num_cols int)
+([][]bool, Random) {
+  rows = [][]bool(num_rows);
+  foreach row_i in 0..num_rows {
+    row = []bool(num_cols);
+    foreach col_i in 0..num_cols {
+      [r, set rand] = RandInt(rand);
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:210]: And also bold this?

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+  }
+  return (rows, rand);
+}
+```
+>>>>
+
+
+Let's break that down:
+
+ * The `rand &Random` parameter became `rand Random`, to take an actual `Random` instance instead of a reference.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:223]: "Again t" I didn't realize at first that we were talking about MakeBoard instead of reiterating how RandInt changed

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 10]
```diff
@@ -0,0 +1,673 @@
...
+Yet there it was.
+
+
+I sat back from my desk, and got up for a walk. After ten minutes, I realized I'd just been staring at a wall, and hadn't actually walked anywhere.
+
+
+It slowly sank in over the next few days.
+
+
+*Using own-lending, we can completely eliminate every single generation check in a program, to get memory safety without borrow checking, reference counting, or tracing garbage collection.*
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: lots of commas, let's try to remove this one.... change ", to get" to "resulting in"

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 11]
```diff
@@ -0,0 +1,673 @@
...
+
+
+That's how I felt in that moment, seeing that zero.
+
+
+Of course, I hadn't discovered anything new. I was discovering for myself the natural consequence of Vale's owned values being [linear](https://en.wikipedia.org/wiki/Substructural_type_system#Linear_type_systems), which is a distant cousin of Rust's affine types. Both are a kind of "[substructural type system](https://en.wikipedia.org/wiki/Substructural_type_system)". Apparently, people have been writing about substructural type systems' abilities for decades! [# There's also a very interesting language named [Austral](https://austral-lang.org/), based on linear types!]
+
+
+With this new understanding, I also considered Rust from a different perspective. It led me to the eventual truth: their memory safety doesn't come from borrow checking, not exactly. It comes from their types being affine. In a way, `&mut` references are just syntactic sugar, [# "Syntactic sugar" is what we call features that just make the code a bit cleaner, but aren't really required to express a certain computation.] and *you can take `&mut` out of Rust and not much would change semantically*.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: Should this be at the end of the line?

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 12]
```diff
@@ -0,0 +1,673 @@
...
+  set x = x + (x / 200096i64);
+  set x = x - (x * 33554432i64);
+  set x = x + (x / 134217728i64);
+  set rand.num = x;
+  return x;
+}
+```
+>>>>
+
+[# E: I used `inout` even though `borrow` would make more sense, because I wanted to equate it to Swift's/C#'s inout, and so the user wouldnt accidentally think it worked like a `&mut`... not sure I succeeded, especially since I mention `&mut` below. `&mut` is part of the type, this is part of the parameter/variable. Think its worth making it clearer, or is it fine?]
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: Seems like you inlined this, so yeah it's fine

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 13]
```diff
@@ -0,0 +1,673 @@
...
+
+# Addressing the Drawbacks
+
+Are there any ways we can address those downsides without sacrificing the benefits?
+
+
+In other words, is there a way we can have linear types _and_ shared mutability?
+
+
+Adding borrow checking partially helps with drawback #1 [# Partially because borrow checking lets us have temporary references, not persistent ones. The borrow checker has a hard time putting references into data structures.] but it still has drawbacks #2, #3, and #4. Rust's `Rc` and `RefCell` helps with those, but they're incompatible with linear types' benefits [# We can't put a linear type into an `Rc` because `Rc` throws away its contents when the last alias disappears.], and have their own drawbacks (like forcing heap allocation and not having [higher RAII](https://verdagon.dev/blog/higher-raii-7drl)).
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:507]: "Rust's `Rc` and `RefCell` helps with those, but they're incompatible with linear types' benefits"

Maybe make this clearer that these things help Rust but won't help Vale. Like "Rust's RC and RefCell helps with those drawbacks, but we can't use those approaches in Vale. They are incompatible with linear types's benefits and additionally have their own drawbacks (etc...)

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 14]
```diff
@@ -0,0 +1,673 @@
...
+With this, we can address a lot of the drawbacks:
+
+ 1. We can have as many references to data as we want, though there is still only one owner.
+ 1. We don't need to restructure our program to accommodate them.
+ 1. We can use patterns like intrusive data structures and graphs. [# In Vale, we would still have all nodes owned by a central collection, but we could still have references between nodes just fine.]
+ 1. We can use observers, back-references, dependency references, callbacks, delegates, whatever we want.
+ 1. We still don't have data races, because Vale changes the generations of any objects that cross region boundaries (unless we safely avoid that with [isolates](/blog/zero-cost-borrowing-regions-part-2-isolates)).
+
+
+Generational references seem to occupy a sweet spot because they allow objects to be linear (they're owned by only one place), yet allow shared mutability in a way that doesn't artificially extend the lifetime of the object.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:523]: "in a way that doesn't artificially extend the lifetime of the object."
Where did this come from? First time you've mentioned it... maybe add "like blah blah blah does" or something

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 15]
```diff
@@ -0,0 +1,673 @@
...
+
+
+Since we're combining linear types with aliasing, we'll call this the *linear-aliasing model*.
+
+
+The mark of a good language is when its features compose well and don't conflict with each other. It seems we're on the right track with this model, as it composes linear types with shared mutability pretty well.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:529]: "pretty well."
You're just restating your definition, maybe define what 'pretty well' means here, like "cohesively and without adding a ton of edge cases"

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 16]
```diff
@@ -0,0 +1,673 @@
...
+
+
+<slice/>
+
+# Matches Made in Heaven
+
+A pleasant surprise, we're back to where we started!
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:536]: I might be dumb... but we started here?

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 17]
```diff
@@ -0,0 +1,673 @@
...
+
+
+And it made me wonder, what does this mean for [regions](/blog/zero-cost-borrowing-regions-overview), Vale's other mechanism for eliminating generation checks?
+
+
+Then I realized, they fit _perfectly_ together. We can use regions to look at a function's inputs in a zero-cost way, and use this "linear style" for any new data the function creates.
+
+I tried this out in a little [roguelike game](https://github.com/Verdagon/RegionsBenchmarks/blob/main/tiny-roguelike/roguelike.vale). The vast majority of generation checks were eliminated by regions, and the rest were eliminated by this new linear style.
+
+[# E: Should I go into more detail about this balance, or save it for a future article?]
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: Depends how long it would be, I'd say you're at length for this article so save it, but if it's really short and not worth its own article then add it here

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 18]
```diff
@@ -0,0 +1,673 @@
...
+
+
+<slice/>
+
+# Conclusion and Next Steps
+
+For those of us that like to squeeze every little last drop of performance out of our programs, this is a pretty amazing realization.
+
+
+But ironically, that's a small minority. For the vast majority of Vale programs, generational references aren't a noticeable source of overhead, and none of this will matter. And that's a good thing; one of Vale's main strengths is that you can just make your program, and things will just work and will be fast.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd:569]: remove last "just"

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 19]
```diff
@@ -0,0 +1,673 @@
...
+
+
+However, this does open up Vale to some new domains. Since we can program in a way that doesnt require generation checks, that means that Vale could be used in new niches, like embedded systems, operating system kernels, and safety-critical software. That's pretty exciting.
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: Feels like you need one final sentence of hype, like "this is an awesome discovery and we're can't wait to think of what else it could lead to" or something

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 20]
```diff
@@ -0,0 +1,673 @@
...
+<center>
+  <a href="https://github.com/sponsors/ValeLang" class="donate-button">
+     <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true" class="octicon octicon-heart icon-sponsor mr-1 color-fg-sponsors">
+        <path fill-rule="evenodd" d="M4.25 2.5c-1.336 0-2.75 1.164-2.75 3 0 2.15 1.58 4.144 3.365 5.682A20.565 20.565 0 008 13.393a20.561 20.561 0 003.135-2.211C12.92 9.644 14.5 7.65 14.5 5.5c0-1.836-1.414-3-2.75-3-1.373 0-2.609.986-3.029 2.456a.75.75 0 01-1.442 0C6.859 3.486 5.623 2.5 4.25 2.5zM8 14.25l-.345.666-.002-.001-.006-.003-.018-.01a7.643 7.643 0 01-.31-.17 22.075 22.075 0 01-3.434-2.414C2.045 10.731 0 8.35 0 5.5 0 2.836 2.086 1 4.25 1 5.797 1 7.153 1.802 8 3.02 8.847 1.802 10.203 1 11.75 1 13.914 1 16 2.836 16 5.5c0 2.85-2.045 5.231-3.885 6.818a22.08 22.08 0 01-3.744 2.584l-.018.01-.006.003h-.002L8 14.25zm0 0l.345.666a.752.752 0 01-.69 0L8 14.25z"></path>
+     </svg>
+     Sponsor us on GitHub!
+  </a>
+</center>
+
+With your support, we can bring weird memory safety paradigms to programmers worldwide!
```
Comment (SquirrelThief) [inline src/blog/0.3/linear-types-borrowing.vmd]: could add a subscript describing "weird" here as "new and super cutting edge" or something like that

File: src/blog/0.3/linear-types-borrowing.vmd (hunk 1 above)
Comment (SquirrelThief) [reply src/blog/0.3/linear-types-borrowing.vmd]: I'm no scholar and these concepts have been bouncing around awhile individually, but I still felt a glimmer of what Champollion felt when he connected two existing ideas to unlock something completely new! Vale showed me that we can use linear types to completely eliminate Vale's memory safety costs in a new paradigm called the linear-aliasing model.

File: src/blog/0.3/linear-types-borrowing.vmd [hunk 21]
```diff
@@ -0,0 +1,673 @@
...
+////
+...to this:
+
+```
+// Returns a 2D array of random bools.
+func MakeBoard(
+   rand Random,
+   num_rows int,
+   num_cols int)
+([][]bool, Random) {
```
Comment (Verdagon) [inline src/blog/0.3/linear-types-borrowing.vmd:205]: perhaps add // 1
